keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,450 @@
+components losing track of the health of other node-local components. For
+example, networking components may unnecessarily route to unhealthy endpoints
+if they're relying on stale information from the API Server.
+
+The new API should return Pods readiness that is computed by Kubelet even if
+they are not reported back to kube-API but reflect the current Pod’s state
+locally.
+
+It is worth noting that a node may lose connectivity to the Control Plane even
+if the Control Plane is healthy. In that case, the Node itself would eventually
+be marked as unhealthy. We believe it is still acceptable for node-local
+components to get node-local Pod readiness information as this would not have
+any impact on other Nodes in the cluster.
+
+#### Kubelet restarts issue
+
+There is a Kubelet restart
+[issue](https://github.com/kubernetes/kubernetes/issues/100277) that could
+impact this API. After restart Kubelet marks all Pods as NotReady and then
+starts the init loop checking the actual pod's status. Kubelet should return
+the pod's readiness information only after the initial loop is finished
+otherwise the API should return gRPC error `FAILED_PRECONDITION`. This
+API needs to be reliable and returned data should reflect the actual state of
+the Pods. We don't want this API to return partial data.
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `status` in Kubelet, listening on a unix socket at `/var/lib/Kubelet/status/Kubelet.sock`. The endpoint will be versioned.
+The gRPC Service will expose 3 methods serving local Pods statuses data:
+
+```protobuf
+service Status {
+    // ListPodConditions returns a of List of PodConditions
+    rpc ListPodConditions(PodConditionsListRequest) returns (PodConditionsListResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsListResponse) {}
+    // GetPodConditions returns a PodReadiness for given pod's UID
+    rpc GetPodConditions(PodConditionsRequest) returns (PodReadiness) {}
+}
+
+// PodCondition aligns with v1.PodCondition.
+message PodCondition {
+    PodConditionType Type = 1;
+    ConditionStatus Status = 2;
+    Timestamp LastProbeTime = 3;
+    Timestamp LastTransitionTime = 4;
+    string Reason = 5;
+    string Message = 6;
+}
+
+// PodConditionType aligns with v1.PodConditionType
+enum PodConditionType {
+    ContainersReady = 0;
+	Initialized = 1;
+	Ready = 2;
+	PodScheduled = 3;
+	DisruptionTarget = 4;
+}
+
+// ConditionStatus aligns with v1.ConditionStatus
+enum ConditionStatus {
+    True = 0;
+	False = 1;
+	Unknown = 2;
+}
+
+// PodReadiness returns a Pod details and list of status Conditions with deletion info.
+message PodReadiness {
+    string podUID = 1;
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated PodCondition conditions = 5;
```

Comment:
please fix the indentation in this section. Hard to read

Reply:
Updated, thanks!